test/parse.jl

```diff
@@ -771,19 +771,26 @@ macro iter()
 end
 end
 let ex = expand(M16096, :(@iter))
-    @test isa(ex, Expr) && ex.head === :body
+    @test isa(ex, Expr) && ex.head === :thunk
 end
 let ex = expand(Main, :($M16096.@iter))
-    @test isa(ex, Expr) && ex.head === :body
+    @test isa(ex, Expr) && ex.head === :thunk
 end
 let ex = expand(@__MODULE__, :(@M16096.iter))
-    @test isa(ex, Expr) && ex.head === :body
+    @test isa(ex, Expr) && ex.head === :thunk
     @test !isdefined(M16096, :foo16096)
-    @test eval(@__MODULE__, ex) === nothing
+    local_foo16096 = eval(@__MODULE__, ex)
+    @test local_foo16096(2.0) == 1
     @test !@isdefined foo16096
-    @test isdefined(M16096, :foo16096)
+    @test !@isdefined it
+    @test !isdefined(M16096, :foo16096)
+    @test !isdefined(M16096, :it)
+    @test_broken typeof(local_foo16096).name.module === M16096
+    @test_broken typeof(local_foo16096).name.mt.module === M16096
+    @test_broken getfield(M16096, typeof(local_foo16096).name.mt.name) === local_foo16096
+    @test_broken !isdefined(@__MODULE__, typeof(local_foo16096).name.mt.name)
```

Comment:
Hygiene means that you either need `esc` (so hygiene rules don't apply), or it'll gensym the name (so you'll define `#1#foo#`). I suppose either case is actually OK to me. I was thinking it might make more sense to not be creating a global with a gensym name – but the goal of this PR was only to restore the v0.6 behavior for these cases, not propose changes.